src/crab/ebpf_domain.cpp

```diff
@@ -174,12 +199,77 @@ bool ebpf_domain_t::operator==(const ebpf_domain_t& other) const {
     return stack == other.stack && m_inv <= other.m_inv && other.m_inv <= m_inv;
 }
 
+// Check whether a given type value is within the range of a given type variable's
+// value.
+bool ebpf_domain_t::variable_has_type(variable_t type_variable, int type) const {
+    crab::interval_t interval = m_inv.eval_interval(type_variable);
+    if (interval.is_top())
+        return true;
+    return (interval.lb().number().value_or(INT_MIN) <= type) &&
+           (interval.ub().number().value_or(INT_MAX) >= type);
+}
+
+void ebpf_domain_t::add_extra_invariant(std::map<crab::variable_t, crab::interval_t>& extra_invariants,
+                                        variable_t type_variable,
+                                        int type,
+                                        variable_t v,
+                                        ebpf_domain_t& other) const {
+    bool has_type = variable_has_type(type_variable, type);
+    bool other_has_type = other.variable_has_type(type_variable, type);
+
+    // If type is contained in exactly one of type1 or type2,
+    // we need to remember the value.
+    if (has_type && !other_has_type)
+        extra_invariants.emplace(v, m_inv.eval_interval(v));
+    else if (!has_type && other_has_type)
+        extra_invariants.emplace(v, other.m_inv.eval_interval(v));
+}
+
 void ebpf_domain_t::operator|=(ebpf_domain_t&& other) {
     if (is_bottom()) {
         *this = other;
         return;
     }
+    if (other.is_bottom()) {
+        return;
+    }
+
+    // Some variables are type-specific.  Type-specific variables
+    // for a register can exist in the domain whenever the associated
+    // type value is present in the register's types interval (and the
+    // value is not Top), and are absent otherwise.
+    //
+    // If a type value is legal in exactly one of the two domains, a
+    // normal join operation would remove any type-specific variables
+    // from the resulting merged domain since absence from the other
+    // would be interpreted to mean Top.
+    //
+    // However, when the type value is not present in one domain, any
+    // any type-specific variables for that type are instead to be
+    // interpreted as Bottom, so we want to preserve the values of any
```

Comment:
This is the part I don't understand. I assume that `top` means that we don't know anything about the type of a variable. So if in one join operand `op1` we know the type of variable and in the other join operand `op2` we don't know anything about it (i.e., it's `top`) then we cannot say that the result of the join is whatever type we have in `op1`.

Reply:
> I assume that `top` means that we don't know anything about the type of a variable. 

`top` means we don't know anything about the _value_ of a variable (where type is one of the variables).

> So if in one join operand `op1` we know the type of variable and in the other join operand `op2` we don't know anything about it (i.e., it's `top`) then we cannot say that the result of the join is whatever type we have in `op1`.

If in one join operand we know the _value_ of a variable and in the other join operand the _value_ of the variable is not relevant because the variable only applies to a given type which is not present, then we can say that the result of the join is whatever _value_ we have in `op1`.

Example:
* r1.type=stack, r1.stack_offset=100
* r1.type=packet, r1.packet_offset=4

Joined result would be 
r1.type={stack,packet}, r1.stack_offset=100, r1.packet_offset=4

Joining the two values for r1.type results in the union of values.
Joining the offsets results in the union of offsets.  Before splitting the variables in this PR it would have been:
* r1.]type=stack, r1.offset=100
* r1.type=packet, r1.offset=4
joined result:
* r1.type={stack,packet}, r1.offset=[4,100]

So splitting the variables just replaces the r1.offset[4,100] into the more specific r1.stack_offset=100, r1.packet_offset=4 but logically the join operation is a union.

Reply:
This is very helpful. Now I understand some of the @elazarg comments. You want to keep track of implications of the form "if register R has type=T then R.T_offset has value ...". 
 
We need to be careful because the representation of the abstract domain is designed to be a conjunction of constraints, and we have now conjunction of implications.